### ISSUE TITLE
Entity Tag System

### ISSUE THREAD

**Issue by joshuaskelly** [Interrupt/delverengine#129]
# Summary
This feature introduces a tag system for entities that can be used to make decisions.

## Use Cases

1. A breakable that only breaks when hit with a weapon with a specific tag. E.g. A boulder is only broken by a mattock. The mattock would be a weapon with the tag `rock-breaker`.
2. Triggers only fire when the encroaching entity has a specific tag. E.g. A special door only opens if the player is poisoned.

## Proposed JSON
```json
{
  "id": "MyEntity",
  "tags": ["poisoned", "broken", "unique"]
}
```

## Proposed Data Structure
```java
public interface TagCollection {
    void add(String value);
    void add(TagCollection values);
    
    void remove(String value);
    void remove(TagCollection values);
    
    void toggle(String value);
    void toggle(TagCollection values);
    
    void contains(String value);
    void containsAny(TagCollection values);
    void containsAll(TagCollection values);
}
```

**Comment by PythooonUser** [comment]
I could imagine that passing a `TagCollection` to those methods would not be *easy*?! How about just passing an array of strings instead, e.g. `tags.remove(["foo", "bar"])`.

**Comment by joshuaskelly** [comment]
Why would that be difficult? It would look like:

```java
Trigger t;
Player p;
p.tags.remove(t.tags);
```
_Example of a trigger that removes a set of tags_

**Comment by evrimoztamur** [comment]
@PythooonUser I agree with @joshuaskelly as initialisation would be done mainly in the data files. How about a generic collection which _also_ adds method overloads for `Array<E>` and `E[]` though?

```java
public interface TagCollection<E> { 
    void add(E value);
    void add(Array<E> value);
    void add(E[] value);
}
```

**Comment by joshuaskelly** [comment]
I'd also like to present a pretty easy view in the editor. Which typically is a comma separate list of strings.

**Comment by PythooonUser** [comment]
That mock example makes sense. I thought of a non-cached use-case where you would just test for some tags, and not creating a collection first etc.

**Comment by joshuaskelly** [comment]
@evrimoztamur I don't know what being generic gets us?

**Comment by evrimoztamur** [comment]
@joshuaskelly because Java has generics and generics are cool 😎

**Comment by joshuaskelly** [comment]
I think it would over complicate the editor experience.

**Comment by evrimoztamur** [comment]
Completely forgot about the editor, you're right!
